lib/vm/src/threadconditions.rs

```diff
@@ -0,0 +1,108 @@
+use std::collections::HashMap;
+use std::sync::{Arc, LockResult, Mutex, MutexGuard};
+use std::thread::{current, park, park_timeout, Thread};
+
+#[derive(Hash, Eq, PartialEq, Clone, Copy, Debug)]
+struct NotifyLocation {
+    pub address: u32,
+}
+
+#[derive(Debug)]
+struct NotifyWaiter {
+    pub thread: Thread,
+    pub notified: bool,
+}
+#[derive(Debug, Default)]
+struct NotifyMap {
+    pub map: HashMap<NotifyLocation, Vec<NotifyWaiter>>,
+}
+
+/// HashMap of Waiters for the Thread/Notify opcodes
+#[derive(Debug)]
+pub struct ThreadConditions {
+    inner: Arc<Mutex<NotifyMap>>, // The Hasmap with the Notify for the Notify/wait opcodes
+}
```

Comment:
Have we taken a look on dashmap? it might be useful here (instead of locking)

https://github.com/xacrimon/dashmap

Reply:
Yeah, this lock will probably be pretty hot in multi-threaded workloads.

We should also look into using `fnv` or https://github.com/rust-lang/rustc-hash for a small and fast hash function, no need to use a secure hash here.

Reply:
done and done